lib/telemetry_registry.ex

```diff
@@ -0,0 +1,253 @@
+  """
+  @type event() :: :telemetry_registry.event()
+
+  @typedoc """
+  An event definition is composed of an event, description, measurements description, and metadata description.
+  """
+  @type event_definition() :: :telemetry_registry.event_definition()
+
+  @typedoc """
+  A description of what the event represents and when it is emitted.
+  """
+  @type event_description() :: :telemetry_registry.event_description()
+
+  @typedoc """
+  A string representation of the measurements emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include a comment on which unit a value is in. The objective
+  is to inform users.
+  """
+  @type event_measurements() :: :telemetry_registry.event_measurements()
+
+  @typedoc """
+  A string representation of the metadata emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include comments or example values. The objective
+  is to inform users what is available.
+  """
+  @type event_metadata() :: :telemetry_registry.event_metadata()
+
+  @typedoc """
+  A map of event definition meta for an event containing the event, measurements, and metadata descriptions
+  if the event was declared with an event definition. Otherwise, this value will be an empty map.
+  """
+  @type event_meta() :: :telemetry_registry.event_meta()
+
+  @typedoc """
+  A list of spannable events known to the registry in the format of `{event_prefix, event_suffixes}`. For
+  example, given events `[:my_app, :request, :start], [:my_app, :request, :stop], [:my_app, :request, :exception]`
+  a span can be created from the `:start` -> `:stop` or the `:start` -> `:exception` events. These are aggregated
+  as a spannable event `{[:my_app, :request], [:start, :stop, :exception]}`.
+  """
+  @type spannable_event() :: :telemetry_registry.spannable_event()
+
+
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__), only: [format_telemetry_docs: 1, telemetry_event: 1]
+
+      Module.register_attribute(__MODULE__, :telemetry_event,
+        accumulate: true,
+        persist: true
+      )
+
+      Module.register_attribute(__MODULE__, :telemetry_docs,
+        accumulate: true,
+        persist: true
+      )
+    end
+  end
+
+  @doc """
+  Declares a telemetry event. Accepts a telemetry event definition or telemetry event name. A full
+  definition is required for complete documentation and is highly encouraged.
+  """
+  defmacro telemetry_event(event) do
+    docstring =
+      case event do
+        event when is_list(event) ->
+          """
+          * `#{inspect(event)}`
+
+          """
+
+        {:%{}, _, event} ->
+          """
+          * `#{inspect(event[:event])}`
+            * Description: #{event[:description]}
+            * Measurements: `#{event[:measurements]}`
+            * Metadata: `#{event[:metadata]}`
+
+          """
```

Comment:
This assumes that docs will always be the Markdown. I would say that this isn't something that library like Telemetry should assume.

Reply:
It is totally ok for the Elixir `telemetry_registry` bindings to assume this. If you have something else, then you simply don't call this function. :)

Reply:
I think, that there should be no docs by default, and instead the documentation generator should take care about it.

Reply:
We talked about this in the last meeting. ExDoc is not going to hardcode Telemetry into it and am extensible mechanism is likely years away, since the focus of the ExDoc team is on Erlang integration. So it is either this or nothing for the foreseeable feature. And I would hate having to declare my events in three places (on execute, as an attribute, and as docs).

Reply:
As noted, we reached this conclusion last week. If we need to change it down the road to add other formatters, it's just an implementation detail to make it configurable and leave markdown as the default.